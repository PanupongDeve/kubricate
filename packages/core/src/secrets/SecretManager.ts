import type { BaseLoader } from './loaders/BaseLoader.js';
import type { BaseProvider } from './providers/BaseProvider.js';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface SecretOptions<NewSecret extends string, Loader extends keyof any, Provider extends keyof any> {
  name: NewSecret;
  loader?: Loader;
  provider?: Provider;
}

/**
 * SecretManager is a type-safe registry for managing secret providers and their secrets.
 *
 * - Register provider definitions (with type-safe config).
 * - Register named provider instances based on those definitions.
 * - Add secrets referencing the registered providers.
 */
export class SecretManager<
  /**
   * Loader instances that have been registered.
   */
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  LoaderInstances extends Record<string, string> = {},
  // /**
  //  * Definitions of allowed provider types.
  //  * Each definition includes a unique `name` and its `config` structure.
  //  */
  // ProviderDefinitions extends ProviderBase,
  /**
   * Instances of providers that have been registered.
   * Keys are provider names, values are typically unique string identifiers.
   */
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  ProviderInstances extends Record<string, string> = {},
  /**
   * Secret entries added to the registry.
   * Keys are secret names, values are their associated string identifiers.
   */
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  SecretEntries extends Record<string, string> = {},
> {
  /**
   * Internal runtime storage for secret values (not type-safe).
   * Intended for use in actual secret resolution or access.
   */
  private _secrets: Record<string, string> = {};
  private _providers: Record<string, string> = {};
  private _loaders: Record<string, string> = {};

  constructor() {}

  /**
   * Registers a new provider instance using a valid provider name
   *
   * @param provider - The unique name of the provider (e.g., 'Kubernetes.Secret').
   * @param options - Configuration specific to the provider type.
   * @returns A SecretManager instance with the provider added.
   */
  addProvider<NewProvider extends string>(provider: NewProvider, options: BaseProvider) {
    this._providers[provider] = '';
    console.log(`Provider ${provider} added with options:`, options);
    return this as SecretManager<LoaderInstances, ProviderInstances & Record<NewProvider, string>, SecretEntries>;
  }

  /**
   * Sets the default provider for the SecretManager.
   * This provider will be used if no specific provider is specified when adding a secret.
   *
   * Providers support multiple instances, so this is a way to set a default.
   *
   * @param provider - The unique name of the provider (e.g., 'Kubernetes.Secret').
   * @returns A SecretManager instance with the provider added.
   */
  setDefaultProvider(provider: keyof ProviderInstances) {
    console.log(`Default provider set to: ${String(provider)}`);
    return this as SecretManager<LoaderInstances, ProviderInstances, SecretEntries>;
  }

  /**
   * Adds a new loader instance using a valid loader name
   *
   * @param loader - The unique name of the loader (e.g., 'EnvLoader').
   * @param options - Configuration specific to the loader type.
   * @returns A SecretManager instance with the loader added.
   */

  addLoader<NewLoader extends string>(loader: NewLoader, options: BaseLoader) {
    console.log(`Loader ${loader} added with options:`, options);
    return this as SecretManager<LoaderInstances & Record<NewLoader, string>, ProviderInstances, SecretEntries>;
  }

  /**
   * Sets the default loader for the SecretManager.
   * This loader will be used if no specific loader is specified when adding a secret.
   *
   * Loaders support multiple instances, so this is a way to set a default.
   *
   * @param loader - The unique name of the loader (e.g., 'EnvLoader').
   * @returns A SecretManager instance with the loader added.
   */
  setDefaultLoader(loader: keyof LoaderInstances) {
    console.log(`Default loader set to: ${String(loader)}`);
    return this as SecretManager<LoaderInstances, ProviderInstances, SecretEntries>;
  }

  /**
   * Adds a new secret to the registry and links it to an existing provider.
   *
   * @param options - Object with:
   *   - `provider`: Key of a registered provider instance.
   *   - `name`: Name of the new secret.
   * @returns A new SecretManager instance with the secret added.
   */
  addSecret<NewSecret extends string>(options: SecretOptions<NewSecret, keyof LoaderInstances, keyof ProviderInstances>) {
    this._secrets[options.name] = '';
    return this as SecretManager<LoaderInstances, ProviderInstances, SecretEntries & Record<NewSecret, string>>;
  }
}
