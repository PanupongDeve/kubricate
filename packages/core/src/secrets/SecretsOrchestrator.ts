import type { PreparedEffect } from './providers/BaseProvider.js';
import { collectSecretManagers, validateSecretManagers, prepareSecretEffects } from './manager.js';
import type { KubricateConfig, BaseLogger } from '../types.js';
import type { SecretManager, SecretOptions } from './SecretManager.js';
import type { EffectsOptions } from './manager.js';

export class SecretsOrchestrator {
  constructor(
    private config: KubricateConfig,
    private effectOptions: EffectsOptions,
    private logger: BaseLogger
  ) { }

  async validate(): Promise<void> {
    this.logger.info('Collecting secret managers...');
    const managers = collectSecretManagers(this.config);
    this.logger.debug(`Found ${Object.keys(managers).length} secret managers`);

    this.logger.info('Validating secret managers...');
    await validateSecretManagers(managers, this.effectOptions);
    this.logger.debug('Secret managers validated successfully');
  }

  async prepare(): Promise<PreparedEffect[]> {
    this.logger.debug('Preparing secret effects...');
    const managers = collectSecretManagers(this.config);
    this.logger.debug(`Preparing secrets for ${Object.keys(managers).length} managers`);
    return prepareSecretEffects(managers, this.effectOptions);
  }

  injectSecretsToProviders(): void {
    const managers = collectSecretManagers(this.config);
    this.logger.debug(`Injecting secrets for ${Object.keys(managers).length} secret managers`);

    for (const { name, stackName, secretManager } of Object.values(managers)) {
      this.logger.debug(`Injecting secrets for stack: ${stackName}, manager: ${name}`);
      this.injectStackSecrets(stackName, secretManager);
    }

    this.logger.debug('All secrets injected into providers');
  }

  private injectStackSecrets(stackName: string, secretManager: SecretManager): void {
    const secrets = secretManager.getSecrets();
    this.logger.debug(`Stack '${stackName}' has ${Object.keys(secrets).length} secrets`);

    for (const [providerId, provider] of Object.entries(secretManager.getProviders())) {
      this.logger.debug(`Checking provider '${providerId}' for injection support`);
      if ('setSecrets' in provider && typeof provider.setSecrets === 'function') {
        const filteredSecrets = this.filterSecretsByProvider(secrets, providerId, secretManager);
        this.logger.debug(`Injecting ${Object.keys(filteredSecrets).length} secrets into provider '${providerId}'`);
        provider.setSecrets(filteredSecrets);
      }
    }
  }

  private filterSecretsByProvider(
    secrets: Record<string, SecretOptions>,
    providerId: string,
    secretManager: SecretManager
  ): Record<string, SecretOptions> {
    const result: Record<string, SecretOptions> = {};
    const defaultProvider = secretManager.getDefaultProvider();
    this.logger.debug(`Filtering secrets for provider '${providerId}' (default: ${defaultProvider})`);

    for (const [key, entry] of Object.entries(secrets)) {
      const entryProvider = entry.provider ?? defaultProvider;
      if (entryProvider === providerId) {
        result[key] = entry;
      }
    }

    this.logger.debug(`Filtered ${Object.keys(result).length} secrets for provider '${providerId}'`);
    return result;
  }
}
