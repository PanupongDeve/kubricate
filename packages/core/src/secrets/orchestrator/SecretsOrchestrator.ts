import type { BaseLogger } from '../../types.js';
import type { PreparedEffect } from '../providers/BaseProvider.js';
import type { SecretValue } from '../types.js';
import { SecretManagerEngine, type MergedSecretManager } from './SecretManagerEngine.js';
import { SecretMergeEngine, type SecretOrigin } from './SecretMergeEngine.js';
import type { SecretsOrchestratorOptions } from './types.js';

export class SecretsOrchestrator {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private providerCache = new Map<string, any>();

  constructor(private engine: SecretManagerEngine, private logger: BaseLogger) { }

  /**
   * Factory method to create a SecretsOrchestrator instance from options.
   */
  static create(options: SecretsOrchestratorOptions): SecretsOrchestrator {
    const engine = new SecretManagerEngine(options);
    return new SecretsOrchestrator(engine, options.logger);
  }

  /**
   * Validates all secret managers by resolving and loading all declared secrets.
   * Ensures secret loaders are configured correctly and values can be fetched.
   *
   * @remarks
   * Used by: `kubricate secret validate`
   *
   * @description
   * Ensures each secret manager is correctly configured by:
   * - Collecting all managers from config
   * - Resolving loaders
   * - Attempting to load each secret once
   * Logs debug information throughout for traceability.
   *
   * @returns Resolves when all managers validate without throwing.
   *
   * @throws If a loader fails or a secret cannot be loaded.
   */
  async validate(): Promise<void> {
    const managers = this.engine.collect();
    await this.engine.validate(managers);
  }

  /**
   * Generates a list of provider-ready effects by:
   * - Loading all secrets
   * - Merging them across levels (stack, manager, etc.)
   * - Preparing Kubernetes-ready manifests
   * - Merging manifests (e.g., same secret name)
   *
   * @remarks
   * Used by: `kubricate secret apply`
   *
   * @description
   *
   * @returns A list of prepared secret effects
   *
   * @throws If loading or provider preparation fails
   */
  async apply(): Promise<PreparedEffect[]> {
    const managers = this.engine.collect();

    const mergedSecrets = await this.loadAndMergeSecrets(managers);
    const rawEffects = this.prepareFromMergedSecrets(managers, mergedSecrets);
    const finalEffects = this.mergeProviderEffects(rawEffects);

    return finalEffects;
  }

  /**
   * Loads and merges secret values from all secret managers using SecretsMergeEngine.
   */
  private async loadAndMergeSecrets(managers: MergedSecretManager): Promise<Record<string, Record<string, SecretValue>>> {
    const allOrigins: SecretOrigin[] = [];

    for (const entry of Object.values(managers)) {
      const secrets = entry.secretManager.getSecrets();
      const raw = await this.engine.loadSecrets(entry.secretManager, secrets);

      const origins: SecretOrigin[] = Object.entries(raw).map(([key, value]) => ({
        key,
        value,
        source: 'loader',
        providerName: entry.name,
        managerName: entry.name,
        stackName: entry.stackName,
        originPath: [`stack:${entry.stackName}`, `manager:${entry.name}`],
      }));

      allOrigins.push(...origins);
    }

    const mergeEngine = new SecretMergeEngine(this.engine.options.logger, {
      config: this.engine.options.config,
      stackName: 'workspace',
      managerName: 'workspace',
    });

    const merged = mergeEngine.merge(allOrigins);

    // regroup merged into per-manager map
    const result: Record<string, Record<string, SecretValue>> = {};
    for (const entry of Object.values(managers)) {
      const prefix = `${entry.stackName}.${entry.name}`;
      const keys = Object.keys(entry.secretManager.getSecrets());
      result[prefix] = {};
      for (const key of keys) {
        if (merged[key] !== undefined) {
          result[prefix][key] = merged[key];
        }
      }
    }

    return result;
  }

  /**
   * Transforms merged secret values into provider-prepared effects.
   */
  private prepareFromMergedSecrets(
    managers: MergedSecretManager,
    merged: Record<string, Record<string, SecretValue>>
  ): PreparedEffect[] {
    const effects: PreparedEffect[] = [];

    for (const entry of Object.values(managers)) {
      const secrets = entry.secretManager.getSecrets();
      const mergedSecrets = merged[`${entry.stackName}.${entry.name}`];

      for (const key of Object.keys(mergedSecrets)) {
        const provider = entry.secretManager.resolveProvider(secrets[key].provider);
        effects.push(...provider.prepare(key, mergedSecrets[key]));
      }
    }

    return effects;
  }

  /**
   * Deduplicates and merges effects per provider by calling mergeSecrets().
   */
  private mergeProviderEffects(effects: PreparedEffect[]): PreparedEffect[] {
    const grouped = new Map<string, PreparedEffect[]>();

    for (const effect of effects) {
      const providerName = effect.providerName;
      if (!providerName) {
        throw new Error(`[SecretsOrchestrator] Effect "${JSON.stringify(effect)}" has no provider name`);
      }
      const group = grouped.get(providerName) ?? [];
      group.push(effect);
      grouped.set(providerName, group);
    }

    const merged: PreparedEffect[] = [];

    for (const [providerName, group] of grouped.entries()) {
      const provider = this.resolveProviderByName(providerName);
      merged.push(...provider.mergeSecrets(group));
    }

    return merged;
  }

  /**
   * Resolves a provider instance from its name by scanning all SecretManagers.
   * Caches resolved providers for performance.
   *
   * @throws If the provider name is not found in any manager
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private resolveProviderByName(providerName: string): any {
    if (this.providerCache.has(providerName)) {
      return this.providerCache.get(providerName);
    }

    for (const entry of Object.values(this.engine.collect())) {
      const secrets = entry.secretManager.getSecrets();
      this.logger.debug(`[SecretsOrchestrator] Looking for provider "${providerName}" in "${entry.name}"`);
      this.logger.debug(`[SecretsOrchestrator] Found secrets: ${JSON.stringify(secrets)}`);
      for (const { provider } of Object.values(secrets)) {
        if (provider === providerName) {
          const instance = entry.secretManager.resolveProvider(provider);
          this.providerCache.set(providerName, instance);
          return instance;
        }
      }
    }

    throw new Error(`[SecretsOrchestrator] Provider "${providerName}" not found in any registered SecretManager`);
  }
}
