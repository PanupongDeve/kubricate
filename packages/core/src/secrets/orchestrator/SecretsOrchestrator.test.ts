/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { SecretsOrchestrator } from './SecretsOrchestrator.js';
import type { SecretsOrchestratorOptions } from './types.js';
import { SecretManager } from '../SecretManager.js';
import type { PreparedEffect } from '../providers/BaseProvider.js';
// import { InMemoryProvider } from '../providers/InMemoryProvider.js';
// import { InMemoryLoader } from '../loaders/InMemoryLoader.js';

describe('SecretsOrchestrator', () => {
  let mockSecretManager: SecretManager;
  let mockProvider: any;
  let orchestrator: SecretsOrchestrator;
  let mockLogger: any;
  let options: SecretsOrchestratorOptions;

  beforeEach(() => {
    mockLogger = {
      info: vi.fn(),
      warn: vi.fn(),
      debug: vi.fn()
    };

    mockProvider = {
      prepare: vi.fn((name, value) => [{
        providerName: 'kubernetes',
        type: 'kubectl',
        value: {
          kind: 'Secret',
          metadata: { name: 'test-secret', provider: 'kubernetes' },
          data: { [name]: value }
        }
      }]),
      mergeSecrets: vi.fn((effects: PreparedEffect[]) => effects)
    };

    mockSecretManager = {
      getSecrets: vi.fn(() => ({
        DB_PASS: { name: 'DB_PASS', loader: 'env', provider: 'kubernetes' }
      })),
      resolveLoader: vi.fn(() => ({
        getWorkingDir: () => undefined,
        setWorkingDir: vi.fn(),
        load: vi.fn(() => Promise.resolve()),
        get: vi.fn(() => 's3cr3t')
      })) as any,
      resolveProvider: vi.fn(() => mockProvider)
    } as any;

    options = {
      config: {
        stacks: {
          api: {
            getSecretManagers: () => ({
              main: mockSecretManager
            })
          } as any
        },
        secrets: {
          merge: {
            providerLevel: 'autoMerge',
            managerLevel: 'error',
            stackLevel: 'error',
            workspaceLevel: 'error',
          }
        }
      },
      logger: mockLogger,
      effectOptions: { workingDir: '/project' }
    };

    orchestrator = SecretsOrchestrator.create(options);
  });

  it('should validate secrets without errors', async () => {
    await expect(orchestrator.validate()).resolves.not.toThrow();
    expect(mockSecretManager.resolveLoader).toHaveBeenCalled();
  });

  it('should apply and return prepared effects', async () => {
    const effects = await orchestrator.apply();
    expect(effects).toHaveLength(1);
    expect(effects[0].value.kind).toBe('Secret');
    expect(mockProvider.prepare).toHaveBeenCalled();
    expect(mockProvider.mergeSecrets).toHaveBeenCalled();
  });

  it('should cache provider in resolveProviderByName()', () => {
    const provider = orchestrator['resolveProviderByName']('kubernetes');
    expect(provider).toBe(mockProvider);
    const cached = orchestrator['providerCache'].get('kubernetes');
    expect(cached).toBe(mockProvider);
  });

  it('should throw if provider name is not found', () => {
    expect(() => orchestrator['resolveProviderByName']('not-exist')).toThrowError(
      '[SecretsOrchestrator] Provider "not-exist" not found in any registered SecretManager'
    );
  });
});


describe('SecretsOrchestrator Multi-Level Merge Strategy', () => {
  let mockLogger: any;
  let mockProvider: any;
  let mockSecretManager: (name: string, secrets: Record<string, any>) => SecretManager;
  let orchestrator: SecretsOrchestrator;

  beforeEach(() => {
    mockLogger = {
      info: vi.fn(),
      warn: vi.fn(),
      debug: vi.fn(),
    };

    mockProvider = {
      prepare: vi.fn((name, value) => [{
        providerName: 'kubernetes',
        type: 'kubectl',
        value: {
          kind: 'Secret',
          metadata: { name: 'merged-secret', provider: 'kubernetes' },
          data: { [name]: value },
        }
      }]),
      mergeSecrets: vi.fn((effects: PreparedEffect[]) => effects),
    };

    mockSecretManager = (name: string, secrets: Record<string, any>): SecretManager => ({
      getSecrets: vi.fn(() => secrets),
      resolveLoader: vi.fn(() => ({
        getWorkingDir: () => undefined,
        setWorkingDir: vi.fn(),
        load: vi.fn(() => Promise.resolve()),
        get: vi.fn((key) => secrets[key].value),
      })),
      resolveProvider: vi.fn(() => mockProvider),
    }) as any;
  });

  it('snapshots multiple secrets from the same manager', async () => {
    const stacks = {
      stack1: {
        getSecretManagers: () => ({
          svc: mockSecretManager('svc', {
            DB_USER: { loader: 'env', provider: 'kubernetes', value: 'admin' },
            DB_PASS: { loader: 'env', provider: 'kubernetes', value: 's3cr3t' }
          }),
        })
      }
    };

    const options: SecretsOrchestratorOptions = {
      logger: mockLogger,
      effectOptions: {},
      config: {
        stacks: stacks as any,
        secrets: {
          merge: {
            providerLevel: 'autoMerge',
            managerLevel: 'autoMerge',
            stackLevel: 'autoMerge',
            workspaceLevel: 'autoMerge',
          }
        }
      }
    };

    orchestrator = SecretsOrchestrator.create(options);
    const effects = await orchestrator.apply();

    expect(effects).toMatchSnapshot();
  });

  it('throws on stackLevel conflict when strategy is "error"', async () => {
    const stacks = {
      stack1: {
        getSecretManagers: () => ({
          svc1: mockSecretManager('svc1', {
            SHARED_KEY: { loader: 'env', provider: 'kubernetes', value: 'one' },
          }),
          svc2: mockSecretManager('svc2', {
            SHARED_KEY: { loader: 'env', provider: 'kubernetes', value: 'two' },
          }),
        })
      }
    };
  
    const options: SecretsOrchestratorOptions = {
      logger: mockLogger,
      effectOptions: {},
      config: {
        stacks: stacks as any,
        secrets: {
          merge: {
            providerLevel: 'autoMerge',
            managerLevel: 'autoMerge',
            stackLevel: 'error',
            workspaceLevel: 'error',
          }
        }
      }
    };
  
    orchestrator = SecretsOrchestrator.create(options);
  
    await expect(orchestrator.apply()).rejects.toThrowError(
      /\[merge:error:stackLevel] Duplicate key "SHARED_KEY"/
    );
  });
  

});


describe('SecretsOrchestrator Advanced Merge Tests', () => {
  let mockLogger: any;
  let mockProvider: any;
  let mockSecretManager: (name: string, secrets: Record<string, any>) => SecretManager;
  let orchestrator: SecretsOrchestrator;

  beforeEach(() => {
    mockLogger = {
      info: vi.fn(),
      warn: vi.fn(),
      debug: vi.fn(),
    };

    mockProvider = {
      prepare: vi.fn((name, value) => [{
        providerName: 'kubernetes',
        type: 'kubectl',
        value: {
          kind: 'Secret',
          metadata: { name: 'test', provider: 'kubernetes' },
          data: { [name]: JSON.stringify(value) },
        }
      }]),
      mergeSecrets: vi.fn((effects: PreparedEffect[]) => effects),
    };

    mockSecretManager = (name: string, secrets: Record<string, any>): SecretManager => ({
      getSecrets: vi.fn(() => secrets),
      resolveLoader: vi.fn(() => ({
        getWorkingDir: () => undefined,
        setWorkingDir: vi.fn(),
        load: vi.fn(() => Promise.resolve()),
        get: vi.fn((key) => secrets[key].value),
      })),
      resolveProvider: vi.fn(() => mockProvider),
    }) as any;
  });

  it('includes correct SecretOrigin metadata and snapshots effect', async () => {
    const stacks = {
      stack1: {
        getSecretManagers: () => ({
          svc: mockSecretManager('svc', {
            API_KEY: { loader: 'env', provider: 'kubernetes', value: '12345' }
          }),
        })
      }
    };

    const options: SecretsOrchestratorOptions = {
      logger: mockLogger,
      effectOptions: {},
      config: {
        stacks: stacks as any,
        secrets: {
          merge: {
            providerLevel: 'autoMerge',
            managerLevel: 'autoMerge',
            stackLevel: 'autoMerge',
            workspaceLevel: 'autoMerge',
          }
        }
      }
    };

    orchestrator = SecretsOrchestrator.create(options);
    const effects = await orchestrator.apply();

    expect(effects).toMatchSnapshot();
    expect(mockProvider.prepare).toHaveBeenCalledWith('API_KEY', '12345');
  });
});


// describe('SecretsOrchestrator providerLevel: error', () => {
//   let orchestrator: SecretsOrchestrator;
//   let mockLogger: any;

//   beforeEach(() => {
//     mockLogger = {
//       info: vi.fn(),
//       warn: vi.fn(),
//       debug: vi.fn(),
//     };
//     // mockLogger = {
//     //   info: console.log,
//     //   warn: console.warn,
//     //   debug: console.debug,
//     // };
//   });

//   it('should throw on providerLevel conflict with strategy "error"', async () => {
//     const secretManager = new SecretManager()
//       .addLoader('InMemoryLoader', new InMemoryLoader({
//         my_app_key: 'my_app_key_value',
//         my_app_key_2: 'my_app_key_2_value',
//       }))
//       .addProvider(
//         'InMemoryProvider1',
//         new InMemoryProvider({
//           name: 'secret-application',
//         })
//       )
//       .addProvider(
//         'InMemoryProvider2',
//         new InMemoryProvider({
//           name: 'secret-application',
//         })
//       )
//       // .addProvider(
//       //   'ImagePullSecretProvider',
//       //   new ImagePullSecretProvider({
//       //     name: 'secret-application',
//       //   })
//       // )
//       .setDefaultProvider('InMemoryProvider1')
//       .addSecret({ name: 'my_app_key', provider: 'InMemoryProvider1' })
//       .addSecret({ name: 'my_app_key_2', provider: 'InMemoryProvider2' })
//       // .addSecret({ name: 'my_app_key_3', provider: 'ImagePullSecretProvider' })
//       .build();

//     const stacks = {
//       app: {
//         getSecretManagers: () => ({
//           default: secretManager,
//         })
//       }
//     };

//     const options: SecretsOrchestratorOptions = {
//       logger: mockLogger,
//       effectOptions: {},
//       config: {
//         stacks: stacks as any,
//         secrets: {
//           merge: {
//             providerLevel: 'error',
//             managerLevel: 'error',
//             stackLevel: 'error',
//             workspaceLevel: 'error',
//           }
//         }
//       }
//     };

//     orchestrator = SecretsOrchestrator.create(options);

//     await expect(orchestrator.apply()).rejects.toThrowError(
//       /\[merge:error:providerLevel\]/
//     );
//   });
// });

