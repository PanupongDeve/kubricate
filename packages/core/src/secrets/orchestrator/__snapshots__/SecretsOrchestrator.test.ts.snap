// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SecretsOrchestrator Advanced Merge Tests > includes correct SecretOrigin metadata and snapshots effect 1`] = `
[
  {
    "type": "kubectl",
    "value": {
      "data": {
        "API_KEY": ""12345"",
      },
      "kind": "Secret",
      "metadata": {
        "name": "test",
        "provider": "kubernetes",
      },
    },
  },
]
`;

exports[`SecretsOrchestrator Multi-Level Merge Strategy > snapshots merge metadata and conflict trace with warn strategy 1`] = `
[
  {
    "type": "kubectl",
    "value": {
      "data": {
        "SHARED_KEY": "one",
      },
      "kind": "Secret",
      "metadata": {
        "name": "merged-secret",
        "provider": "kubernetes",
      },
    },
  },
  {
    "type": "kubectl",
    "value": {
      "data": {
        "SHARED_KEY": "one",
      },
      "kind": "Secret",
      "metadata": {
        "name": "merged-secret",
        "provider": "kubernetes",
      },
    },
  },
]
`;

exports[`SecretsOrchestrator Multi-Level Merge Strategy > snapshots merge metadata and conflict trace with warn strategy 2`] = `
[
  "[merge:warn:stackLevel] Duplicate key "SHARED_KEY" detected. Keeping first.",
]
`;

exports[`SecretsOrchestrator Multi-Level Merge Strategy > snapshots multiple secrets from the same manager 1`] = `
[
  {
    "type": "kubectl",
    "value": {
      "data": {
        "DB_USER": "admin",
      },
      "kind": "Secret",
      "metadata": {
        "name": "merged-secret",
        "provider": "kubernetes",
      },
    },
  },
  {
    "type": "kubectl",
    "value": {
      "data": {
        "DB_PASS": "s3cr3t",
      },
      "kind": "Secret",
      "metadata": {
        "name": "merged-secret",
        "provider": "kubernetes",
      },
    },
  },
]
`;
